sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueriesSuite.scala

```diff
@@ -0,0 +1,355 @@
+    val subquery2 = ScalarSubquery(testRelation.where('a > 1).select('b.as("b_1")).select('b_1))
+    // Despite lacking a Project node, `subquery3` is mergeable with the result of merging
+    // `subquery1` and `subquery2`
+    val subquery3 = ScalarSubquery(testRelation.select('a.as("a_2")).where('a_2 > 1).select('a_2))
+    val subquery4 = ScalarSubquery(
+      testRelation.select('a.as("a_2"), 'b).where('a_2 > 1).select('b.as("b_2")))
+    val originalQuery = testRelation
+      .select(
+        subquery1,
+        subquery2,
+        subquery3,
+        subquery4)
+
+    val mergedSubquery = testRelation
+      .select('a, 'b, 'c)
+      .where('a > 1)
+      .select('a, 'b, 'c)
+      .select('a, 'b)
+      .select(CreateNamedStruct(Seq(
+        Literal("a"), 'a,
+        Literal("b_1"), 'b
+      )).as("mergedValue"))
+    val analyzedMergedSubquery = mergedSubquery.analyze
+    val correctAnswer = WithCTE(
+      testRelation
+        .select(
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1),
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1)),
+      Seq(CTERelationDef(analyzedMergedSubquery, 0)))
+
+    comparePlans(Optimize.execute(originalQuery.analyze), correctAnswer.analyze)
+  }
+
+  test("Merging subqueries with aggregate filters") {
+    val subquery1 = ScalarSubquery(
+      testRelation.having('b)(max('a).as("max_a"))(max('a) > 1))
+    val subquery2 = ScalarSubquery(
+      testRelation.having('b)(sum('a).as("sum_a"))(max('a) > 1))
+    val originalQuery = testRelation.select(
+      subquery1,
+      subquery2)
+
+    val mergedSubquery = testRelation
+      .having('b)(
+        max('a).as("max_a"),
+        sum('a).as("sum_a"))('max_a > 1)
+      .select(
+        'max_a,
+        'sum_a)
+      .select(CreateNamedStruct(Seq(
+        Literal("max_a"), 'max_a,
+        Literal("sum_a"), 'sum_a
+      )).as("mergedValue"))
+    val analyzedMergedSubquery = mergedSubquery.analyze
+    val correctAnswer = WithCTE(
+      testRelation
+        .select(
+          extractorExpression(0, analyzedMergedSubquery.output, 0),
+          extractorExpression(0, analyzedMergedSubquery.output, 1)),
+      Seq(CTERelationDef(analyzedMergedSubquery, 0)))
+
+    comparePlans(Optimize.execute(originalQuery.analyze), correctAnswer.analyze)
+  }
+
+  test("Merging subqueries with joins") {
+    val subquery1 = ScalarSubquery(testRelation.as("t1")
+      .join(
+        testRelation.as("t2"),
+        Inner,
+        Some($"t1.b" === $"t2.b"))
+      .select($"t1.a").analyze)
+    val subquery2 = ScalarSubquery(testRelation.as("t1")
+      .select('a.as("a_1"), 'b.as("b_1"), 'c.as("c_1"))
+      .join(
+        testRelation.as("t2").select('a.as("a_2"), 'b.as("b_2"), 'c.as("c_2")),
+        Inner,
+        Some('b_1 === 'b_2))
```

Comment:
Add a test case with different join conditions in subquery1 and subquery2?

Reply:
Added a new test in https://github.com/apache/spark/pull/32298/commits/8457148369b6c8835376146fffaa915493016eae